src/i2p.cpp

```diff
@@ -0,0 +1,432 @@
+        }
+
+        const Reply& lookup_reply =
+            SendRequestAndGetReply(sock, strprintf("NAMING LOOKUP NAME=%s", to.ToStringIP()));
+
+        const std::string& dest = lookup_reply.Get("VALUE");
+
+        const Reply& connect_reply = SendRequestAndGetReply(
+            sock, strprintf("STREAM CONNECT ID=%s DESTINATION=%s SILENT=false", session_id, dest),
+            false);
+
+        const std::string& result = connect_reply.Get("RESULT");
+
+        if (result == "OK") {
+            conn.socket = sock.Release();
+            return true;
+        }
+
+        if (result == "INVALID_ID") {
+            LOCK(m_mutex);
+            Disconnect();
+            throw std::runtime_error("Invalid session id");
+        }
+
+        if (result == "CANT_REACH_PEER" || result == "TIMEOUT") {
+            proxy_error = false;
+        }
+
+        throw std::runtime_error(strprintf("'%s'", connect_reply.full));
+    } catch (const std::runtime_error& e) {
+        Log("Error connecting to %s: %s", to.ToString(), e.what());
+        return false;
+    }
+}
+
+// Private methods
+
+std::string Session::Reply::Get(const std::string& key) const
+{
+    const auto& pos = keys.find(key);
+    if (pos == keys.end() || !pos->second.has_value()) {
+        throw std::runtime_error(
+            strprintf("Missing %s= in the reply to '%s': '%s'", key, request, full));
+    }
+    return pos->second.value();
+}
+
+template <typename... Args>
+void Session::Log(const std::string& fmt, const Args&... args) const
+{
+    LogPrint(BCLog::I2P, "I2P: %s\n", tfm::format(fmt, args...));
+}
+
+Session::Reply Session::SendRequestAndGetReply(const Sock& sock,
+                                               const std::string& request,
+                                               bool check_result_ok) const
+{
+    SendComplete(sock.Get(), request + "\n", MAX_WAIT_FOR_IO, *m_interrupt);
+
+    Reply reply;
+
+
+    // Don't log the full "SESSION CREATE ..." because it contains our private key.
+    reply.request = request.substr(0, 14) == "SESSION CREATE" ? "SESSION CREATE ..." : request;
+
+    // It could take a few minutes for the I2P router to reply as it is querying the I2P network
+    // (when doing name lookup, for example). Notice: `RecvUntilTerminator()` is checking
+    // `m_interrupt` more often, so we would not be stuck here for long if `m_interrupt` is
+    // signaled.
+    static constexpr auto recv_timeout = 3min;
+
+    reply.full = RecvUntilTerminator(sock.Get(), '\n', recv_timeout, *m_interrupt);
+
+    for (const auto& kv : spanparsing::Split(reply.full, ' ')) {
+        const auto& pos = std::find(kv.begin(), kv.end(), '=');
+        if (pos != kv.end()) {
+            reply.keys.emplace(std::string{kv.begin(), pos}, std::string{pos + 1, kv.end()});
+        } else {
+            reply.keys.emplace(std::string{kv.begin(), kv.end()}, std::nullopt);
+        }
+    }
+
+    if (check_result_ok && reply.Get("RESULT") != "OK") {
```

Comment:
The `Reply` class could have a `isOk()` function, what do you think?

Reply:
I think it is not worth to add `isOk()` method because it would be used in just one place - here and `reply.Get("RESULT") != "OK"` is readable enough.